### PR TITLE
Tickets/dm 30325 bryce

### PR DESCRIPTION
**Pull request by jbkalmbach**
Iterating with @connolly on a new `DonutStamp` method

### CODE REVIEW

**Comment by jbkalmbach** [comment]
Link to demo notebook: https://gist.github.com/jbkalmbach/90022d66e44122ce561e429220e5aff7

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 1]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+        lsst.afw.cameraGeom.Camera
+            Camera object for the exposures.
+        """
+
+        if self.cam_name == "LSSTCam":
+            cam = obs_lsst.LsstCam().getCamera()
+        elif self.cam_name == "LSSTComCam":
+            cam = obs_lsst.LsstComCam().getCamera()
+        else:
+            raise ValueError(f"Camera {self.cam_name} is not supported.")
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py:156]: Please help to add the doc string of raised `ValueError`. Thanks!

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 2]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+        Returns
+        -------
+        lsst.afw.cameraGeom.Camera
+            Camera object for the exposures.
+        """
+
+        if self.cam_name == "LSSTCam":
+            cam = obs_lsst.LsstCam().getCamera()
+        elif self.cam_name == "LSSTComCam":
+            cam = obs_lsst.LsstComCam().getCamera()
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: You could consider to return the value immediately. Such as:

```
 if self.cam_name == "LSSTCam":
     return obs_lsst.LsstCam().getCamera()
 elif self.cam_name == "LSSTComCam":
     return obs_lsst.LsstComCam().getCamera()
```

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 3]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+
+        return cam
+
+    def calcFieldXY(self):
+        """
+        Calculate the X, Y field position of the centroid in radians.
+
+        Returns
+        -------
+
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: Please fill in the doc string of "Returns".

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 4]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+        elif self.cam_name == "LSSTComCam":
+            cam = obs_lsst.LsstComCam().getCamera()
+        else:
+            raise ValueError(f"Camera {self.cam_name} is not supported.")
+
+        return cam
+
+    def calcFieldXY(self):
+        """
+        Calculate the X, Y field position of the centroid in radians.
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: It is "in radians" or "in degrees"? You called the `np.degrees()` in return values.

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 5]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+
+        field_xy = self.calcFieldXY()
+
+        self.comp_im.setImg(
+            field_xy,
+            DefocalType(self.defocal_type),
+            self.stamp_im.getImage().getArray(),
         )
+        self.cMask = afwImage.MaskX()
+        self.pMask = afwImage.MaskX()
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: Do you want to consider to rename the attributes to be `self.mask_c` and `self.mask_p` to be consistent with `self.comp_im`, etc.? The similar things for other attributes in this PR.

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 6]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+            Neuman boundary where the derivative of the wavefront is set to
+            zero.
+        maskScalingFactorLocal : float
+            Mask scaling factor (for fast beam) for local correction.
+
+        Returns
+        -------
+        cMask : afwImage.MaskX
+            Non-padded mask for use at the offset planes.
+
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: This empty line is not needed.

File: python/lsst/ts/wep/task/DonutStamp.py [hunk 7]
```diff
@@ -85,9 +107,103 @@ def factory(cls, stamp_im, metadata, index, archive_element=None):
...
+        # 0 flag in mask is part of image that is not donut
+        # 1 flag in mask means it is part of the model donut
+        maskDict = {"BKGRD": 0, "DONUT": 1}
+        dimOfDonut = inst.getDimOfDonutOnSensor()
+
+        self.pMask = afwImage.MaskX(lsst.geom.Extent2I(dimOfDonut), maskDict)
+        self.cMask = afwImage.MaskX(lsst.geom.Extent2I(dimOfDonut), maskDict)
+        self.pMask.array = np.array(self.comp_im.pMask, dtype=np.int32)
+        self.cMask.array = np.array(self.comp_im.cMask, dtype=np.int32)
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamp.py]: Is there any set function in `afwImage.MaskX` can be called to set the value of `array` attribute? I thought this class comes from the C++ code and there should be some set function can be used. But I could be wrong. Please help to check. Thanks!

File: python/lsst/ts/wep/task/DonutStamps.py [hunk 8]
```diff
@@ -88,6 +92,28 @@ def getDetectorNames(self):
...
+        return [stamp.cam_name for stamp in self]
+
+    def getDefocalTypes(self):
+        """
+        Get the defocal type for each stamp.
+
+        Returns
+        -------
+        list [int]
+            Defocal types for each stamp.
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/DonutStamps.py]: Please add the doc string of `int` here. For example, what is the meaning if the return value is `[0, 1]`, etc? Another option is to use the `enum` here, which is more clear.

File: python/lsst/ts/wep/task/EstimateZernikesBase.py [hunk 9]
```diff
@@ -337,6 +342,12 @@ def cutOutStamps(self, exposure, donutCatalog, defocalType):
         stampsMetadata["DET_NAME"] = np.array(
             [detectorName] * len(detectorCatalog), dtype=str
         )
+        stampsMetadata["CAM_NAME"] = np.array(
+            [cameraName] * len(detectorCatalog), dtype=str
+        )
+        stampsMetadata["DFC_TYPE"] = np.array(
+            [defocalType.value] * len(detectorCatalog)
+        )
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesBase.py]: I think there is the duplication of `len(detectorCatalog)` here. You could calculate this value in advance in this function.

File: python/lsst/ts/wep/task/EstimateZernikesFamTask.py [hunk 10]
```diff
@@ -80,6 +80,30 @@ def __init__(self, **kwargs):
...
+    def runQuantum(
+        self,
+        butlerQC: pipeBase.ButlerQuantumContext,
+        inputRefs: pipeBase.InputQuantizedConnection,
+        outputRefs: pipeBase.OutputQuantizedConnection,
+    ):
+        """
+        We implement a runQuantum method to make sure our configured
+        task runs with the instrument required by the pipeline.
+        """
```
Comment (teweitsai) [inline python/lsst/ts/wep/task/EstimateZernikesFamTask.py]: Can you update the doc string to have the "Parameters"? Or this is a function from the upstream and you are just doing the overwriting?

File: tests/task/test_donutStamp.py [hunk 11]
```diff
@@ -53,13 +59,18 @@ def _makeStamps(self, nStamps, stampSize):
         centX = np.arange(nStamps) + 20
         centY = np.arange(nStamps) + 25
         detectorNames = ["R22_S11"] * nStamps
+        camNames = ["LSSTCam"] * nStamps
+        dfcTypes = [DefocalType.Extra.value] * nStamps
+        dfcTypes[: int(nStamps / 2)] = [DefocalType.Intra.value] * int(nStamps / 2)
```
Comment (teweitsai) [inline tests/task/test_donutStamp.py]: I think there is the duplication here: `int(nStamps / 2)`. You could calculate it first and give a sensible name to improve  the readability.

File: tests/task/test_donutStamp.py [hunk 12]
```diff
@@ -87,3 +98,95 @@ def testFactory(self):
...
+
+        donutStamp = DonutStamp(
+            self.testStamps[0],
+            lsst.geom.SpherePoint(0.0, 0.0, lsst.geom.degrees),
+            lsst.geom.Point2D(2047.5, 2001.5),
+            DefocalType.Extra.value,
+            "R22_S11",
+            "LSSTCam",
+        )
+        np.testing.assert_array_almost_equal(donutStamp.calcFieldXY(), (0, 0))
```
Comment (teweitsai) [inline tests/task/test_donutStamp.py:147]: Can you help to test the field position of corner wavefront sensors as well? I am interested to know the values of 8 corner wavefront sensors. Or you could let Chris to try this to see the values are consistent with the `focalplane.txt` he updated in the `phosim_syseng4` or not.

File: tests/task/test_donutStamp.py [hunk 13]
```diff
@@ -87,3 +98,95 @@ def testFactory(self):
...
+        # Check masks after creation
+        donutStamp.makeMasks(inst, "offAxis", 0, 1)
+        self.assertEqual(afwImage.MaskX, type(donutStamp.cMask))
+        self.assertEqual(afwImage.MaskX, type(donutStamp.pMask))
+        self.assertDictEqual(
+            {"BKGRD": 0, "DONUT": 1}, donutStamp.cMask.getMaskPlaneDict()
+        )
+        self.assertDictEqual(
+            {"BKGRD": 0, "DONUT": 1}, donutStamp.pMask.getMaskPlaneDict()
+        )
```
Comment (teweitsai) [inline tests/task/test_donutStamp.py]: I begin to think you could consider to have an enum for `BKGRD` and `DONUT`.

File: tests/task/test_donutStamps.py [hunk 14]
```diff
@@ -56,13 +57,18 @@ def _makeDonutStamps(self, nStamps, stampSize):
         centX = np.arange(nStamps) + 20
         centY = np.arange(nStamps) + 25
         detectorNames = ["R22_S11"] * nStamps
+        camNames = ["LSSTCam"] * nStamps
+        dfcTypes = [DefocalType.Extra.value] * nStamps
+        dfcTypes[: int(nStamps / 2)] = [DefocalType.Intra.value] * int(nStamps / 2)
```
Comment (teweitsai) [inline tests/task/test_donutStamps.py]: Same comment above for `int(nStamps / 2)`.

File: tests/task/test_donutStamps.py [hunk 15]
```diff
@@ -122,6 +128,23 @@ def testGetDetectorNames(self):
...
+    def testGetDefocalTypes(self):
+
+        defocalTypes = self.donutStamps.getDefocalTypes()
+        self.assertListEqual(
+            defocalTypes[: int(self.nStamps / 2)],
+            [DefocalType.Intra.value] * int(self.nStamps / 2),
+        )
+        self.assertListEqual(
+            defocalTypes[int(self.nStamps / 2) :],
+            [DefocalType.Extra.value] * int((self.nStamps - int(self.nStamps / 2))),
```
Comment (teweitsai) [inline tests/task/test_donutStamps.py]: Same comment above for `int(nStamps / 2)`.

**Review by teweitsai — APPROVED** [review]
The update looks good to me. I left some comments in PR. Please consider them before the merge. Thanks!

File: python/lsst/ts/wep/task/DonutStamp.py (hunk 7 above)
Comment (jbkalmbach) [reply python/lsst/ts/wep/task/DonutStamp.py]: Yeah, I will just do this in the opposite order and set the array when originally defining the `Mask` object and then use the included functions to set the new mask plane definitions.

File: tests/task/test_donutStamp.py (hunk 12 above)
Comment (jbkalmbach) [reply tests/task/test_donutStamp.py:147]: Ok, I added a useful test to compare the positions of the corner detectors getting them a different way as a comparison.

File: tests/task/test_donutStamp.py (hunk 13 above)
Comment (jbkalmbach) [reply tests/task/test_donutStamp.py]: I think I will leave this the way it is for now. If we start using this dictionary in other places in the code then we can change how it is defined.

**Pull request merged**